File: freecodecamp/daily-code-challenge/2026-03-21.py
def rotate(mat):
    return [''.join(row[i] for row in reversed(mat)) for i in range(6)]

def is_valid(mat):
    def check(r, c):
        return (mat[r][c] == '1' and mat[r][c+1] == '1' and
                mat[r+1][c] == '1' and mat[r+1][c+1] == '1')
    
    return check(0, 0) and check(0, 4) and check(4, 0)

def decode_qr(qr_code):
    mat = qr_code
    
    # Try all 4 rotations
    for _ in range(4):
        if is_valid(mat):
            break
        mat = rotate(mat)
    
    result = []
    
    for i in range(6):
        for j in range(6):
            # Skip 3 orientation blocks
            if (i < 2 and j < 2) or (i < 2 and j >= 4) or (i >= 4 and j < 2):
                continue
            result.append(mat[i][j])
    
    return ''.join(result)

File: freecodecamp/daily-code-challenge/2026-03-21.js
function rotate(mat) {
  let res = Array.from({ length: 6 }, () => Array(6));
  for (let i = 0; i < 6; i++) {
    for (let j = 0; j < 6; j++) {
      res[j][5 - i] = mat[i][j];
    }
  }
  return res.map(row => row.join(''));
}

function isValid(mat) {
  function check(r, c) {
    return mat[r][c] === '1' && mat[r][c+1] === '1' &&
           mat[r+1][c] === '1' && mat[r+1][c+1] === '1';
  }
  return check(0,0) && check(0,4) && check(4,0);
}

function decodeQr(qrCode) {
  let mat = qrCode;

  for (let k = 0; k < 4; k++) {
    if (isValid(mat)) break;
    mat = rotate(mat);
  }

  let result = [];

  for (let i = 0; i < 6; i++) {
    for (let j = 0; j < 6; j++) {
      if ((i < 2 && j < 2) || (i < 2 && j >= 4) || (i >= 4 && j < 2)) continue;
      result.push(mat[i][j]);
    }
  }

  return result.join('');
}
